core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -112,68 +112,70 @@ final class FiberRefs private (
     val parentFiberRefs = self.fiberRefLocals
     val childFiberRefs  = that.fiberRefLocals
 
+    @tailrec
+    def findAncestor(initialValue: Any)(
+      parentStack: List[StackEntry[?]],
+      parentDepth: Int,
+      childStack: List[StackEntry[?]],
+      childDepth: Int
+    ): Any =
+      parentStack match {
+        case Nil => initialValue // case (Nil, _)
+        case StackEntry(parentFiberId, parentValue, parentVersion) :: parentAncestors =>
+          childStack match {
+            case Nil => initialValue // case (_, Nil)
+            case StackEntry(
+                  childFiberId,
+                  childValue,
+                  childVersion
+                ) :: childAncestors => // case (head0 :: tail0, head1 :: tail1)
+              if (parentFiberId == childFiberId)
+                if (childVersion > parentVersion) parentValue else childValue
+              else if (childDepth > parentDepth)
+                findAncestor(initialValue)(parentStack, parentDepth, childAncestors, childDepth - 1)
+              else if (childDepth < parentDepth)
+                findAncestor(initialValue)(parentAncestors, parentDepth - 1, childStack, childDepth)
+              else
+                findAncestor(initialValue)(parentAncestors, parentDepth - 1, childAncestors, childDepth - 1)
+          }
+      }
+
     val fiberRefLocals0 = childFiberRefs.foldLeft(parentFiberRefs) {
       case (parentFiberRefs, (fiberRef, Value(childStack, childDepth))) =>
-        val ref        = fiberRef.asInstanceOf[FiberRef[Any]]
-        val childValue = childStack.head.value
-
-        if (childStack.head.id == fiberId) {
-          parentFiberRefs
-        } else {
+        if (childStack.head.id == fiberId) parentFiberRefs
+        else {
+          val ref          = fiberRef.asInstanceOf[FiberRef[Any]]
+          val initialValue = ref.initial
+          val childValue   = childStack.head.value
 
-          parentFiberRefs
-            .get(ref)
-            .fold {
-              if (childValue == ref.initial) parentFiberRefs
+          parentFiberRefs.getOrElse(ref, null) match {
+            case null =>
+              if (childValue == initialValue) parentFiberRefs
```

Comment:
I wonder what it means to compare this `childStack.head.value` with `ref.initial` 🤔